cluster/terraform_kubernetes/kube_state_metrics.tf

```diff
@@ -0,0 +1,160 @@
+    name = "kube-system"
+  }
+}
+
+locals {
+  kube_system_ns = data.kubernetes_namespace.kube_system.id != null ? data.kubernetes_namespace.kube_system.id : kubernetes_namespace.kube_system[0].id
+}
+
+resource "kubernetes_cluster_role" "kube_state_metrics" {
+  metadata {
+    name = "kube-state-metrics"
+    labels = {
+      "app.kubernetes.io/component" = "exporter"
+      "app.kubernetes.io/name" = "kube-state-metrics"
+      "app.kubernetes.io/version" = "2.3.0"
+    }
+  }
+
+  rule {
+    api_groups = [""]
+    resources = ["configmaps", "secrets", "nodes", "pods", "services", "resourcequotas", "replicationcontrollers", "limitranges", "persistentvolumeclaims", "persistentvolumes", "namespaces", "endpoints"]
+    verbs = ["list", "watch"]
+  }
+}
+
+resource "kubernetes_cluster_role_binding" "kube_state_metrics" {
+  metadata {
+    name = "kube-state-metrics"
+    labels = {
+      "app.kubernetes.io/component" = "exporter"
+      "app.kubernetes.io/name" = "kube-state-metrics"
+      "app.kubernetes.io/version" = "2.3.0"
+    }
+  }
+
+  role_ref {
+    api_group = "rbac.authorization.k8s.io"
+    kind = "ClusterRole"
+    name = kubernetes_cluster_role.kube_state_metrics.metadata[0].name
+  }
+
+  subject {
+    kind = "ServiceAccount"
+    name = kubernetes_service_account.kube_state_metrics.metadata[0].name
+    namespace = local.kube_system_ns
+  }
+}
+
+resource "kubernetes_service_account" "kube_state_metrics" {
+  metadata {
+    name = "kube-state-metrics"
+    namespace = local.kube_system_ns
+    labels = {
+      "app.kubernetes.io/component" = "exporter"
+      "app.kubernetes.io/name" = "kube-state-metrics"
+      "app.kubernetes.io/version" = "2.3.0"
+    }
+  }
+}
+
+resource "kubernetes_deployment" "kube_state_metrics" {
+  metadata {
+    name = "kube-state-metrics"
+    namespace = local.kube_system_ns
+    labels = {
+      "app.kubernetes.io/component" = "exporter"
+      "app.kubernetes.io/name" = "kube-state-metrics"
+      "app.kubernetes.io/version" = "2.3.0"
+    }
+  }
+
+  spec {
+    replicas = 1
+
+    selector {
+      match_labels = {
+        "app.kubernetes.io/name" = "kube-state-metrics"
+      }
+    }
+
+    template {
+      metadata {
+        labels = {
+          "app.kubernetes.io/component" = "exporter"
+          "app.kubernetes.io/name" = "kube-state-metrics"
+          "app.kubernetes.io/version" = "2.3.0"
+        }
+      }
+
+      spec {
+        automount_service_account_token = true
+        service_account_name = kubernetes_service_account.kube_state_metrics.metadata[0].name
+
+        container {
+          name  = "kube-state-metrics"
+          image = "k8s.gcr.io/kube-state-metrics/kube-state-metrics:v2.3.0"
+
+          port {
+            name = "http-metrics"
+            container_port = 8080
+          }
+
+          port {
+            name = "telemetry"
+            container_port = 8081
+          }
+
+          # Add readiness/liveness probes, resources, etc
```

Comment:
can we add probes and memory/resource limits

Reply:
done